module/zfs/ddt_stats.c

```diff
@@ -138,20 +139,59 @@ ddt_get_dedup_object_stats(spa_t *spa, ddt_object_t *ddo_total)
 		for (ddt_type_t type = 0; type < DDT_TYPES; type++) {
 			for (ddt_class_t class = 0; class < DDT_CLASSES;
 			    class++) {
+				dmu_object_info_t doi;
+				uint64_t cnt;
+				int err;
+
+				/*
+				 * XXX this is a repeat of of the cached stats
+				 *     load in ddt_object_load(). feel like
+				 *     one of these isn't needed anymore, which
+				 *     is it? -- robn, 2024-02-07
+				 */
+
+				err = ddt_object_info(ddt, type, class, &doi);
+				if (err != 0)
+					continue;
+
+				err = ddt_object_count(ddt, type, class, &cnt);
+				if (err != 0)
+					continue;
+
 				ddt_object_t *ddo =
 				    &ddt->ddt_object_stats[type][class];
+
+				ddo->ddo_count = cnt;
+				ddo->ddo_dspace =
+				    doi.doi_physical_blocks_512 << 9;
+				ddo->ddo_mspace = doi.doi_fill_count *
+				    doi.doi_data_block_size;
+
 				ddo_total->ddo_count += ddo->ddo_count;
 				ddo_total->ddo_dspace += ddo->ddo_dspace;
 				ddo_total->ddo_mspace += ddo->ddo_mspace;
 			}
 		}
 	}
 
-	/* ... and compute the averages. */
-	if (ddo_total->ddo_count != 0) {
-		ddo_total->ddo_dspace /= ddo_total->ddo_count;
-		ddo_total->ddo_mspace /= ddo_total->ddo_count;
-	}
+	/*
+	 * XXX this dspace/mspace change in quota is returning raw counts
+	 *     to userspace, rather than averages. at least, wants to be
+	 *     a separate commit, but also think hard about what this is
+	 *     even trying to achieve -- robn, 2024-02-07
+	 */
```

Comment:
It looks to me like this might be a long standing bug and `print_dedup_stats()` has always expected raw counts.  At least that's the only in-tree consumer I see.  Is there another case?

Reply:
Updated comment